records/2024-06-04.manifests-kustomize-cicd.md

```diff
@@ -0,0 +1,134 @@
+# Current Manifests CI/CD
+
+Date: 2024-06-04
+
+## Status
+
+**DRAFT**.
+
+## Related ADRs
+
+For information on how manifest rollouts are done using GitHub ARC Runners please see this ADR:
+[Private EKS Cluster](2024-02-15.private-eks-cluster.md)
+
+## Context
+
+The way the CI/CD works for releases for manifests is not well documented. This ADR outlines the the CI/CD release process for staging and production, the current problems we face with this model, and proposed solutions moving forward.
+
+## Current State
+
+### Staging
+
+#### Individual Application Rollouts
+
+When deploying a new version of the application components (admin/api/document-download etc), the release to staging is kicked off at the application repository level when the docker image is built. The docker image tag is based on the first 7 digits of the github runner sha, and also tags this build as latest. The docker build and push workflow invokes a corresponding workflow in the notification-manifests repository passing the new docker tag to have it deployed to staging.
+
+The manifests rollout workflow receives the new image tag and performs a manual patch of the existing deployment:
+
+```shell
+kubectl set image deployment.apps/api api=$DOCKER_SLUG:$DOCKER_TAG -n=notification-canada-ca 
+```
+
+##### Diagram
+
+![Database activity stream architecture](./diagrams/2024-06-04.manifests-current-cicd/staging-cicd.png)
+
+#### Global Manifests Rollouts
+
+When a global manifests rollout is performed by the merge to main (staging) github workflow, all of the Kustomize code is deployed.
+
+```shell
+notification-manifests/env/staging kubectl apply -k .
+```
+
+### Production
+
+#### Individual Application Rollouts
+
+This concept does not exist for production. We only rollout all of Kustomize.
+
+#### Global Manifests Rollouts
+
+Every night, the Notify PR Bot reads the latest github commit sha's for each repository, and if they've changed, creates a pull request that changes the image tag values in the Kustomization.yaml for production. Once this is merged, all of the kustomize code is deployed at once.
```

Comment:
This is a nightly thing? I thought it happened whenever we made changes?

Reply:
Yes, for each workflow of each component, the docker workflow notifies the PR bot:
https://github.com/cds-snc/notification-api/blob/4ad86fd5ff1c5e4c8345381dbb0aa656b912828e/.github/workflows/docker.yaml#L63

Reply:
My bad, I'll update